### PR TITLE
Update ethereum/tests / Reintroduce VM debug logger with safe guard

### PR DESCRIPTION
**Pull request by holgerd77**
This PR updates `ethereum/tests`.

It will do this two steps:

1. Update to latest [v8.0.5](https://github.com/ethereum/tests/releases/tag/8.0.5) release and see how things go (first CI on PR opening)
2. Update to latest [develop](https://github.com/ethereum/tests) which now also includes the EIP-1559 blockchain tests from [here](https://github.com/ethereum/tests/pull/837)

CI is run with the `type: test all hardforks` label.

### CODE REVIEW

**Comment by holgerd77** [comment]
Calling a contract with an undefined opcode as first byte of the code from here https://github.com/ethereum/tests/pull/836 is currently failing, can be run with `ts-node ./tests/tester --state --fork=Berlin --test='undefinedOpcodeFirstByte'`.

**Comment by holgerd77** [comment]
I will reintroduce the `debug` logger here removed in #1198 in its most naive approach by just adding all debug lines and then commenting them out. Then it is at least possible to do proper debugging in the specific situations one is needing it by simply doing a search/replace on `//debug(`, then do the debugging and after finding the error just discard all the changes with a `git checkout`.

**Comment by ryanio** [comment]
for the debug logger, could you guard it within an if clause to skip if undefined? like `if (process.env.DEBUG) { debug() }` it would be verbose but I know the goal is to not execute the string literal.

**Comment by holgerd77** [comment]
@ryanio is it guaranteed that this doesn't execute the string literal then? Then I can do for sure.

**Comment by holgerd77** [comment]
@ryanio ok, I will do. The current guard will break in the browser though since `process` is not available, will do a slight modification.

**Comment by holgerd77** [comment]
Phew, what an impressively shitty work to re-add all this stuff one had already done months ago. 😬  So happy when I'm through with this.

**Comment by holgerd77** [comment]
Ok, this is side-tracking too much and the bug here is unrelated to the 1559 changes and not really very important and in the VM for a long time. I have opened a dedicated issue which we can handle separately and will add to the skipped tests and continue here.

**Comment by holgerd77** [comment]
London tests are currently all failing, see e.g. `ts-node ./tests/tester --blockchain --fork=London --file='CREATE2_EOF1'`.

This is due to a difference in the naming of the base fee field, in the tests this is called `baseFee` while we are calling it `baseFeePerGas`, have opened a question [here](https://github.com/ethereum/tests/pull/837#issuecomment-848698711) on `ethereum/tests`.

The base fee on our side therefore defaults to `7` and then differs from the base fee provided

**Comment by holgerd77** [comment]
Update: I've now fixed this (at least temporarily) for both state and blockchain tests.

This should now provide a basis to start with the real thing and do fixes on failing tests.

@ryanio @jochem-brouwer I would ask you to really prioritize this, I likely can't continue with this today any more, so that we are getting not too much into delay! Thanks!

**Comment by holgerd77** [comment]
Current failing tests:

## State Tests

Run all with: `ts-node ./tests/tester --state --fork=London`

2 failures (total 150)

- `basefeeExample`, run with `ts-node ./tests/tester --state --fork=London --test='basefeeExample'``
- `eip1559`, run with `ts-node ./tests/tester --state --fork=London --test='eip1559'`

## Blockchain Tests

10 failures (total 1275)

- `badBlocks`, run with `ts-node ./tests/tester --blockchain --fork=London --file='badBlocks'`
- `gasLimit20m`, `ts-node ./tests/tester --blockchain --fork=London --file='gasLimit20m'`
- `gasLimit40m`, `ts-node ./tests/tester --blockchain --fork=London --file='gasLimit40m'`

**Comment by ryanio** [comment]
@holgerd77 thanks for working on this, i can help triage these bugs the rest of today after our call.

i had one more idea for the debug logger guard, if we don't want to make every 1 debug line into 3 with an if statement, i just tried this experiment:

```typescript
const DEBUG = true
const debug = console.debug ?? createDebugLogger('vm:state')

let didRun = false

const evaluate = () => {
    didRun = true
    return 'hello'
}

let debugGuard = (fn: Function) => { 
    if (DEBUG) {
      debug(fn())
    }
}

debugGuard(() => `${evaluate()}`)
console.log('didRun: ', didRun)

didRun = false
debugGuard = () => {} // set to no-op

console.log('---')

debugGuard(() => `${evaluate()}`)
console.log('didRun: ', didRun)
```

returns

```
> ts-node test.ts
hello
didRun:  true
---
didRun:  false
```

so what we could do is wrap all the debug with e.g. debugGuard and execute the string literal in a function that would never be called in non-debug by setting debugGuard to no-op.

**Comment by ryanio** [comment]
Failing state tests should be resolved with https://github.com/ethereumjs/ethereumjs-monorepo/pull/1269/commits/8ba316d9c1153eb2ebd004992554863ff5fd9b70

For the blockchain tests, I described the issues in https://github.com/ethereum/tests/pull/837#issuecomment-849201825

**Comment by ryanio** [comment]
rebased on master, updated ethereum-tests submodule to https://github.com/ethereum/tests/commit/1bb7d823e445388353f8f8f9858c221ffb7aaa37

**Comment by holgerd77** [comment]
Rebased this.

**Comment by ryanio** [comment]
Rebased and updated to ethereum/tests [v9.0.1](https://github.com/ethereum/tests/releases/tag/9.0.1)

**Comment by holgerd77** [comment]
This is slowly getting urgent (will add a respective label) and we should prioritize here to get the tests up and running, since `london` is approaching quickly and the first testnet forks are on the horizon (fork dates likely today I guess?).

So open for anyone to pick up. I've asked Dimitry on the #tooling Discord channel on the state of the `ethereum/tests` repo, was a bit unsure lately since there was so much activity going on.

**Comment by holgerd77** [comment]
Updated this to `ethereum/tests` [v9.0.2](https://github.com/ethereum/tests/releases/tag/9.0.2) (London update, 2021-06-12).

**Comment by holgerd77** [comment]
Failing `london` state tests:

- `intrinsic`: `ts-node ./tests/tester --state --fork=London --test='intrinsic'`
- `lowFeeCap`: `ts-node ./tests/tester --state --fork=London --test='lowFeeCap'`
- `lowGasLimit`: `ts-node ./tests/tester --state --fork=London --test='lowGasLimit'`
- `tipTooHigh`: `ts-node ./tests/tester --state --fork=London --test='tipTooHigh'`

Very welcome to be continued by someone else, need to stop now! 😄

**Comment by ryanio** [comment]
I wanted to dig into these four failing state tests today but ran out of time, I will investigate tomorrow but if someone wants to take a look before I get to it please go ahead, thanks for highlighting the particular ones @holgerd77.

**Comment by holgerd77** [comment]
Rebased this on top of the new config PR #1297

**Comment by holgerd77** [comment]
Ok, I guess this is working now. This needed a fix to include failed tx validation into the test runs (before it was just throwing within the test runner and going out of the test execution context), see 45b9bec.

Ready for review. 😄

File: packages/vm/tests/config.ts [hunk 1]
```diff
@@ -11,11 +11,25 @@ export const DEFAULT_FORK_CONFIG = 'Istanbul'
...
+  // (temporary until fixed 2021-05-26)
+  'CREATE2_EOF1',
+  'CREATE2_FirstByte_loop',
+  'CREATE_EOF1',
+  'CREATE_FirstByte_loop',
+  'refundFF',
+  'refundMax',
+  'refundSSTORE',
+  'CreateTransactionEOF1',
+  'Opcodes_TransactionInit',
```
Comment (ryanio) [inline packages/vm/tests/config.ts:32]:
```suggestion
remove skipped tests
```

Comment (ryanio) [inline packages/vm/tests/config.ts:32]:
```suggestion
```

**Review by ryanio — APPROVED** [review]
lgtm!

**Comment by ryanio** [comment]
@alcuadrado just a note we reintroduced the vm `debug` logger here but with a much more careful approach to guard string interpolation under `if (this.DEBUG)` statements.

Our [benchmarks](http://ethereumjs.github.io/ethereumjs-monorepo/dev/bench/vm/) didn't seem to report a deviation but let us know in a next release if you notice any slow down.

Benchmark [result](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1269/checks?check_run_id=2835440845#step:8:14) from this PR:

```
Block 9422905 x 1,924 ops/sec ±3.19% (91 runs sampled)
Block 9422906 x 1,851 ops/sec ±4.79% (90 runs sampled)
Block 9422907 x 1,749 ops/sec ±9.09% (88 runs sampled)
Block 9422908 x 1,860 ops/sec ±1.05% (93 runs sampled)
Block 9422909 x 1,562 ops/sec ±10.72% (79 runs sampled)
Block 9422910 x 1,788 ops/sec ±1.17% (93 runs sampled)
Block 9422911 x 1,760 ops/sec ±1.22% (94 runs sampled)
Block 9422912 x 1,755 ops/sec ±1.14% (92 runs sampled)
Block 9422913 x 1,766 ops/sec ±1.24% (93 runs sampled)
Block 9422914 x 1,540 ops/sec ±8.80% (84 runs sampled)
```

File: packages/vm/tests/config.ts [hunk 2]
```diff
@@ -17,19 +17,6 @@ export const SKIP_BROKEN = [
...
-  // (temporary until fixed 2021-05-26)
-  'CREATE2_EOF1',
-  'CREATE2_FirstByte_loop',
-  'CREATE_EOF1',
-  'CREATE_FirstByte_loop',
-  'refundFF',
-  'refundMax',
-  'refundSSTORE',
-  'CreateTransactionEOF1',
-  'Opcodes_TransactionInit',
```
Comment (holgerd77) [inline packages/vm/tests/config.ts]: Oh nice, wasn't aware of these skipped tests any more.

**Pull request merged**